### PR TITLE
[java] Mock environment without requiring new constructor

### PR DESCRIPTION
**Pull request by titusfortner**
This addressed Environment Management more like how #95 addresses Driver Management instead of using PowerMockito like #93

### CODE REVIEW

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 1]
```diff
@@ -122,14 +109,52 @@ public void sauceOptions_startWithChrome_startsChrome() {
...
+            sauce.start();
+            fail("Expected a SauceEnvironmentVariablesNotSetException to be thrown");
+        } catch (SauceEnvironmentVariablesNotSetException exception) {
+            assertEquals("Sauce Access Key was not provided", exception.getMessage());
+        }
+    }
+
+    @Test
+    public void startThrowsErrorWithoutAccessKey() {
+        sauce = Mockito.spy(new SauceSession(options, dummyRemoteDriver));
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Mockito.spy' due to existing static import 'org.mockito.Mockito.*'](https://app.codacy.com/app/SauceLabs/simple_sauce/pullRequest?prid=4732268)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/SauceLabs/simple_sauce/pullRequest?prid=4732268&bid=15799507)

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 2]
```diff
@@ -122,14 +109,52 @@ public void sauceOptions_startWithChrome_startsChrome() {
...
+        sauce = new SauceSession(dummyRemoteDriver);
 
         WebDriver driver = sauce.start();
         sauce.stop();
 
         driver.quit();
     }
 
+    @Test
+    public void usesENVAsDefaultUsername() {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/app/SauceLabs/simple_sauce/pullRequest?prid=4732268)

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 3]
```diff
@@ -122,14 +109,52 @@ public void sauceOptions_startWithChrome_startsChrome() {
...
+        sauce = Mockito.spy(new SauceSession(options, dummyRemoteDriver));
+        doReturn("test-name").when(sauce).getEnvironmentVariable("SAUCE_USERNAME");
+    }
+
+    @Test
+    public void usesENVAsDefaultAccessKey() {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/app/SauceLabs/simple_sauce/pullRequest?prid=4732268)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/SauceLabs/simple_sauce/pullRequest?prid=4732268&bid=15799507)

File: java/src/main/java/com/saucelabs/simplesauce/SauceSession.java [hunk 4]
```diff
@@ -35,43 +35,45 @@ public MutableCapabilities getSauceOptionsCapability(){
...
-        this.environmentManager = environmentManager;
     }
 
     public WebDriver start() {
-        mutableCapabilities = appendSauceCapabilities();
-        setBrowserSpecificCapabilities(sauceOptions.getBrowserName());
-        currentSessionCapabilities = setRemoteDriverCapabilities(mutableCapabilities);
-        tryToCreateRemoteWebDriver(sauceDataCenter);
-        return webDriver;
+        if (getSauceUsername() == null) {
```
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/simplesauce/SauceSession.java]: This `start()` method got hairy really quick with the if else clauses. Any reason you don't want to just keep this login in the corresponding methods where they existed here?
```
        mutableCapabilities.setCapability("username", getUserName());	        mutableCapabilities.setCapability("username", getSauceUsername());
        mutableCapabilities.setCapability("accessKey", getAccessKey());
```

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 5]
```diff
@@ -49,16 +44,14 @@ public void startSession_defaultConfig_usWestDataCenter() {
 
     @Test
     public void getUserName_usernameSetInEnvironmentVariable_returnsValue() {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: Hmmm, something doesn't seem right here. If the user name is set in Env variables, why would it `assertNotEquals` to the expected value which is "test-name"? Shouldn't it be the opposite?

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 6]
```diff
@@ -49,16 +44,14 @@ public void startSession_defaultConfig_usWestDataCenter() {
...
+        String actualUserName = sauce.getSauceUsername();
+        assertNotEquals("test-name",actualUserName);
     }
 
     @Test
     public void getAccessKey_keySetInEnvironmentVariable_returnsValue() {
-        when(dummyEnvironmentManager.getEnvironmentVariable("SAUCE_ACCESS_KEY")).thenReturn("accessKey");
-        String actualAccessKey = sauce.getAccessKey();
-        assertNotEquals("", actualAccessKey);
+        String actualAccessKey = sauce.getSauceAccessKey();
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: Same thing here, I'd expect it to return "accesskey" and **be equal**

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 7]
```diff
@@ -122,14 +108,56 @@ public void sauceOptions_startWithChrome_startsChrome() {
...
+
+        assertEquals("accesskey", sauce.getSauceAccessKey());
+    }
+
+    @Test
+    public void startThrowsErrorWithoutUsername() {
+        sauce = spy(new SauceSession(options, dummyRemoteDriver));
+        doReturn(null).when(sauce).getEnvironmentVariable("SAUCE_ACCESS_KEY");
+
+        try {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: The same thing might be accomplished by adding this annotation to the test and removing the try catch.
`@Test(expected = SauceEnvironmentVariablesNotSetException.class)`
I know it won't test the error message, but is that a necessary test?

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java [hunk 8]
```diff
@@ -122,14 +108,56 @@ public void sauceOptions_startWithChrome_startsChrome() {
...
+            sauce.start();
+            fail("Expected a SauceEnvironmentVariablesNotSetException to be thrown");
+        } catch (SauceEnvironmentVariablesNotSetException exception) {
+            assertEquals("Sauce Access Key was not provided", exception.getMessage());
+        }
+    }
+
+    @Test
+    public void startThrowsErrorWithoutAccessKey() {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java:132]: Same as the user name test?

**Review by nadvolod — COMMENTED** [review]
Left a few comments and queries

File: java/src/main/java/com/saucelabs/simplesauce/SauceSession.java (hunk 4 above)
Comment (titusfortner) [reply java/src/main/java/com/saucelabs/simplesauce/SauceSession.java]: yeah, for security reasons it needs to go into the URL
Current implementation effectively does the evaluation in the start method, but puts it in the getter, and I don't think a getter should throw an exception. If there isn't a username set it should return null

Now, what we probably should do is to have the public setter/getter be `username` / `accessKey` and have `getSauceUsername` and `getSauceAccessKey` be private methods that throw this error. Does that sound good?

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java (hunk 5 above)
Comment (titusfortner) [reply java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: yeah, this is a dumb test. I delete it when merging in #94 when we can evaluate that ENV are getting used by evaluating the public `getSauceURL` where it actually matters.

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java (hunk 6 above)
Comment (titusfortner) [reply java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: yeah, it's taking advantage of the fact that my "real" `ENV["SAUCE_ACCESS_KEY"]` is not "accesskey" which I agree is kind of silly test.

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java (hunk 7 above)
Comment (titusfortner) [reply java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: No, error message not necessary. Yours is a better pattern, I'll update

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java (hunk 5 above)
Comment (titusfortner) [reply java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: actually, this is redundant with the tests
`usesENVAsDefaultUsername` and `usesENVAsDefaultAccessKey`

File: java/src/main/java/com/saucelabs/simplesauce/SauceSession.java (hunk 4 above)
Comment (titusfortner) [reply java/src/main/java/com/saucelabs/simplesauce/SauceSession.java]: ok, I implemented what I'm thinking in the latest commit

File: java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java (hunk 5 above)
Comment (titusfortner) [reply java/src/test/java/com/saucelabs/simplesauce/SauceSessionTest.java]: of course with the change I made above, that makes those 2 tests not work until we have #94 implemented. Merge that one in and I'll rebase this so we can un-ignore these tests. :)

**Comment by nadvolod** [comment]
@titusfortner looking good here besides the merge conflicts. Feel free to resolve and merge in

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/simple_sauce/pullRequest?prid=4732268&bid=15799507)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/simple_sauce/pullRequest?prid=4732268&bid=15799507)

**Pull request merged**